core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -120,48 +123,65 @@ public SnapshotsService(Settings settings, ClusterService clusterService, IndexN
     }
 
     /**
+     * Retrieves list of snapshot ids that are present in a repository
+     *
+     * @param repositoryName repository name
+     * @return list of snapshot ids
+     */
+    public List<SnapshotId> snapshotIds(final String repositoryName) {
+        Repository repository = repositoriesService.repository(repositoryName);
+        assert repository != null; // should only be called once we've validated the repository exists
+        return repository.snapshots();
+    }
+
+    /**
      * Retrieves snapshot from repository
      *
-     * @param snapshotId snapshot id
+     * @param repositoryName  repository name
+     * @param snapshotId      snapshot id
      * @return snapshot
      * @throws SnapshotMissingException if snapshot is not found
      */
-    public SnapshotInfo snapshot(SnapshotId snapshotId) {
-        validate(snapshotId);
-        List<SnapshotsInProgress.Entry> entries = currentSnapshots(snapshotId.getRepository(), new String[]{snapshotId.getSnapshot()});
+    public SnapshotInfo snapshot(final String repositoryName, final SnapshotId snapshotId) {
+        List<SnapshotsInProgress.Entry> entries = currentSnapshots(repositoryName, Arrays.asList(snapshotId.getName()));
         if (!entries.isEmpty()) {
             return inProgressSnapshot(entries.iterator().next());
         }
-        return repositoriesService.repository(snapshotId.getRepository()).readSnapshot(snapshotId);
+        return repositoriesService.repository(repositoryName).readSnapshot(snapshotId);
     }
 
     /**
      * Returns a list of snapshots from repository sorted by snapshot creation date
      *
      * @param repositoryName repository name
+     * @param snapshotIds       snapshots for which to fetch snapshot information
+     * @param ignoreUnavailable if true, snapshots that could not be read will only be logged with a warning,
+     *                          if false, they will throw an error
      * @return list of snapshots
      */
-    public List<SnapshotInfo> snapshots(String repositoryName, boolean ignoreUnavailable) {
-        Set<SnapshotInfo> snapshotSet = new HashSet<>();
-        List<SnapshotsInProgress.Entry> entries = currentSnapshots(repositoryName, null);
+    public List<SnapshotInfo> snapshots(final String repositoryName, List<SnapshotId> snapshotIds, final boolean ignoreUnavailable) {
+        final Set<SnapshotInfo> snapshotSet = new HashSet<>();
+        final List<SnapshotId> snapshotIdsToIterate = new ArrayList<>(snapshotIds);
+        final List<SnapshotsInProgress.Entry> entries =
+            currentSnapshots(repositoryName, snapshotIds.stream().map(SnapshotId::getName).collect(Collectors.toList()));
         for (SnapshotsInProgress.Entry entry : entries) {
             snapshotSet.add(inProgressSnapshot(entry));
+            snapshotIdsToIterate.remove(entry.snapshot().getSnapshotId());
```

Comment:
So, this is not exactly true.  In `TransportGetSnapshotsAction`, `TransportSnapshotsStatusAction`, and in `SnapshotsService#snapshots`, we first look at the current in progress snapshots, then at the repository.  This is necessary to avoid looking at the repository for snapshots that don't exist in it (thereby throwing exceptions).  Also, in each of these, the idea is that *if* there are two snapshots of the same name, the one from the repository should override it, given that we know the one "in-progress" is soon to be deleted from the cluster state due to the naming conflict.  However, `TransportSnapshotsStatusAction` did not do this, due to the line of checking if the snapshot is in currentSnapshotIds and if so, skips over it.  Instead I changed it as can be found in the latest commit here: https://github.com/elastic/elasticsearch/pull/18228/commits/a17039c86d7d9bfd10f77bbb15ca1b7a43496226#diff-9426f24b722fd1c3664ad17ead551f83L211

This leaves `deleteSnapshot` to have the only "inconsistent" behavior... but I believe it doesn't make sense to look first in current, only to know we have to look then into the repository and if found in the repository, discard the one in current, for the purposes of deletion.